contracts/epochs/FlowEpoch.cdc

```diff
@@ -406,6 +541,92 @@ pub contract FlowEpoch {
             FlowEpoch.account.load<Bool>(from: /storage/flowAutomaticRewardsEnabled)
             FlowEpoch.account.save(enabled, to: /storage/flowAutomaticRewardsEnabled)
         }
+        
+        /// Ends the currently active epoch and starts a new one with the provided configuration.
+        /// This function saves the recovery epoch metadata in storage containing the full specification for the new epoch.
+        /// This meta data will be emitted in the EpochRecover service event in the next heart beat interval.
+        /// This function is used within sporks to recover the network from Epoch Fallback Mode (EFM).
+        pub fun recoverEpoch(
+            randomSource: String,
+            startView: UInt64,
+            stakingEndView: UInt64,
+            endView: UInt64,
+            collectorClusters: [[String]],
+            clusterQCVoteData: [FlowClusterQC.ClusterQCVoteData],
+            dkgPubKeys: [String],
+            nodeIDs: [String])
+        {
+            pre {
+                FlowEpoch.isValidPhaseConfiguration(stakingEndView-startView+1, FlowEpoch.configurableMetadata.numViewsInDKGPhase, endView-startView+1):
+                    "Invalid startView, stakingEndView, and endView configuration"
+            }
+
+            if FlowEpoch.currentEpochPhase == EpochPhase.STAKINGAUCTION {
+                // Since we are resetting the epoch, we do not need to
+                // start epoch setup also. We only need to end the staking auction
+                FlowEpoch.borrowStakingAdmin().endStakingAuction()
+            } else {
+                // force reset the QC and DKG
+                FlowEpoch.borrowClusterQCAdmin().forceStopVoting()
+                FlowEpoch.borrowDKGAdmin().forceEndDKG()
+            }
+
+            // Construct the identity table for the recovery epoch
+            let nodes: [FlowIDTableStaking.NodeInfo] = []
+            
+            for nodeID in nodeIDs {
+                /// @TODO: if node is SN ensure weight > 0
+                /// throw exception, or add with 0 weight
+                nodes.append(FlowIDTableStaking.NodeInfo(nodeID))
+            }
+
+            // Compute the target end time for the next epoch
+            let timingConfig = FlowEpoch.getEpochTimingConfig()
+            let proposedTargetDuration = timingConfig.duration
+            let proposedTargetEndTime = timingConfig.getTargetEndTimeForEpoch(FlowEpoch.proposedEpochCounter())
+
+            // Create new Epoch metadata for the next epoch
+            // with the new values
+            let newEpochMetadata = EpochMetadata(
+                counter: FlowEpoch.currentEpochCounter + 1,
```

Comment:
I am not sure this is correct. During EFM, the Protocol State formally remains in the same epoch, which is regularly being extended until we reach an EpochRecovery event. However, the Epoch Smart contract does not know this and behaves as if the epochs where just progressing on the happy path. So very likely, the Epoch smart contract and the Protocol State have a different notion of what the "next" epoch counter should be. 

Same suggestion as before:
Even if it is ok to compute the epoch counter like this, I would like a comment that explains why this is safe and which assumptions we are making (i.e. under which conditions this code would need to change).

Reply:
It is safe to use the epoch counter in the smart contract because the counter is not incremented when we enter EFM state. Som, despite adding EpochExtensions the counter will not be incremented until the manual intervention via epoch recover transaction. Added a comment https://github.com/onflow/flow-core-contracts/pull/420/commits/4c0dcf7bbfd903fb6a08e8bd61520542e8d70135